### DISCUSSION TITLE
Charging schedule

### DISCUSSION THREAD

**Discussion by felixnurse** [Hyundai-Kia-Connect/kia_uvo#369] [Ideas]
Kia has a bug where the charging schedule is forgotten. Not sure if it's in the car or the service. I have an automation to catch when the schedule is forgotten so I know to reset it before plugging it back in and wasting money.
It seems like you are able to add services for most elements of the app.
Please, please, please could you add a service to setup schedules so it can be synced with off peak electricity tariffs?

**Comment by cdnninja** [comment]
Start charge is already supported as a service.   Just create a new automation to start charge at the time or condition you like. Should work but let us know!

**Reply by felixnurse** [reply]
Yes - you're correct - the schedule can be managed via HASS automations.

However, the automation cannot catch when I plug the car in, and it starts charging straight away due to the timings of when the integration checks in with the car.
It's not a huge deal as I am able to catch when the schedule is forgotten, and then go into the app. It would just be good to be able to then set a new schedule from the integration as well.

**Reply by richy58729** [reply]
I have dabbled with two automations to start charging in low tariff periods and to stop when in a high tariff period, but:
- Whenever the car is plugged in, it seems like it automatically sends that state to Kia, and I saw the plugged-in state also in HA (IIRC after updating from the cache), so that's at least a good starting point.
- However, when the car itself doesn't have a charging schedule, it will just start charging automatically.
- The automation to stop charging when in high tariff periods checks every ten seconds and would then stop the charging because it's a high tariff period.
- Something like twenty seconds later the car would start charging again because there is no charging schedule.
- The automation to stop charging when in high tariff periods checks every ten seconds and would then stop the charging because it's a high tariff period.
- ... Rinse and repeat...

So I just disabled those two automations because I don't know if it would have adverse effects on my car, my charger, etc. with perhaps high costs involved when charging is constantly started, stopped, started, stopped etc. and something breaks because of it....

The charging schedule (and the climate control schedule) in the car and Kia Connect app can be seen in the raw data in HA, but that unfortunately is read-only. I now have climate control scheduling running in HA completely with a combination of some Lovelace cards, helper variables, scripts and automations, and it works much better than from the Kia Connect app IMHO. Maybe one or two nice-to-haves, but overall, the UX is just great (and more possibilities, more flexibility, etc.).

Today I was considering that somehow I should be able to change the charging schedule from the Kia Connect app and see what is communicated to and from the API, so it could be made into a new service from this integration. If someone has ideas how to set that up, I would be more than willing to help with delivering the communication between the app and the API to @cdnninja (or another contributor to this repo), so he/she can add yet another service to the integration (like the set_charge_limits service a short while ago, which is awesome, thanks again!).

I was even considering perhaps contributing even more to this repo, but I also need a little bit of help there, as I now have a working HA debugging environment in Visual Studio Code in combination with WSL and Docker, but also getting HACS to work in that environment (and this integration) I haven't succeeded in doing yet, so pointers in the right direction would be appreciated. Then all I have to do is getting up to speed with coding in Python (I have experience in PowerShell, Visual Basic for Applications and a tiny little bit in JavaScript and C#, but not yet in Python)...

**Reply by cdnninja** [reply]
This is awesome!   Love that you are interested in helping out.  A key reason items like this haven't moved is because I don't own an electric car.  I really is easy to add a service stub out to this repo.  The hard part is understanding the API and mapping the values.  Adding sensors is also really easy if we know the path. Happy to point you in the right direction on this as to what areas you need to edit to add this.  Would that work?  

As for the visual studio side best route is a https://developers.home-assistant.io/docs/development_environment/
@fuatakgun  wrote me steps on how to use this for a custom integration like this awhile back.  I will see if I can find those steps.  

Other less ideal route is using your prod instance and editing the files in visual studios via samba but that isn't recommended.

**Reply by cdnninja** [reply]
Here is a few example PRs that show the basics:
Adding a service start to finish to a single region: https://github.com/fuatakgun/kia_uvo/pull/169/files
How an example of a EU service instead of the above region: https://github.com/fuatakgun/kia_uvo/pull/357/files
Adding a sensor with existing data.  You simple replicate this PR but with the correct fields, icons and data path: https://github.com/fuatakgun/kia_uvo/pull/244/files

**Reply by richy58729** [reply]
> This is awesome! Love that you are interested in helping out. A key reason items like this haven't moved is because I don't own an electric car. I really is easy to add a service stub out to this repo. The hard part is understanding the API and mapping the values. Adding sensors is also really easy if we know the path. Happy to point you in the right direction on this as to what areas you need to edit to add this. Would that work?
> 
> As for the visual studio side best route is a https://developers.home-assistant.io/docs/development_environment/ @fuatakgun wrote me steps on how to use this for a custom integration like this awhile back. I will see if I can find those steps.
> 
> Other less ideal route is using your prod instance and editing the files in visual studios via samba but that isn't recommended.

As it is also for my own benefit and I like to code in my spare time, it actually is a no-brainer. And yeah, it was already clear to me it looks like there isn't someone contributing to this repo with an electric car, so happy to help out there.

The link in your comment is exactly the link I used to get the development environment up-and-running. It would be great if you can find those steps back.

**Reply by richy58729** [reply]
> Here is a few example PRs that show the basics: Adding a service start to finish to a single region: https://github.com/fuatakgun/kia_uvo/pull/169/files How an example of a EU service instead of the above region: https://github.com/fuatakgun/kia_uvo/pull/357/files Adding a sensor with existing data. You simple replicate this PR but with the correct fields, icons and data path: https://github.com/fuatakgun/kia_uvo/pull/244/files

Thanks for the links. I'll look into them.

**Reply by cdnninja** [reply]
@kunalkhosla Are you able to  chime in on dev container usage? I see you added the dev container config.
